src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.Tests.cs

```diff
@@ -79,6 +79,28 @@ public void Constructor_NoEntryInsertion_WritesNothing()
             Assert.Equal(0, archiveStream.Length);
         }
 
+        [Fact]
+        public void Write_To_UnseekableStream()
+        {
+            MemoryStream stream = new MemoryStream();
+            using WrappedStream wrapped = new WrappedStream(stream, canRead: true, canWrite: true, canSeek: false);
+
+            using (TarWriter writer = new TarWriter(wrapped, leaveOpen: true))
+            {
+                PaxTarEntry paxEntry = new PaxTarEntry(TarEntryType.RegularFile, "file.txt");
+                writer.WriteEntry(paxEntry);
+            } // The final records should get written, and the length should not be set because position cannot be read
+
+            stream.Seek(0, SeekOrigin.Begin); // Rewind the base stream (wrapped cannot be rewinded)
+
+            using TarReader reader = new TarReader(wrapped);
+            TarEntry entry = reader.GetNextEntry();
+            Assert.NotNull(entry);
+            Assert.Equal(TarFormat.Pax, reader.Format);
+            Assert.Equal(TarEntryType.RegularFile, entry.EntryType);
+            Assert.Null(reader.GetNextEntry());
+        }
+
         [Fact]
```

Comment:
We have a test that verifies the behavior of the `TarEntry.DataStream` when reading an unseekable stream:

https://github.com/dotnet/runtime/blob/2389815a1bc810b5b3c39586b570cbaae2f0f3a5/src/libraries/System.Formats.Tar/tests/TarReader/TarReader.GetNextEntry.Tests.cs#L165

But that test wasn't creating an archive using `TarWriter` against an unseekable stream, hence why this bug wasn't caught before. That's why I'm also adding this extra test to write to a generic unseekable stream.